File: fabric-audit-agent/core/detectors/report.js
import { DEFAULT_CONFIG } from '../config.js';

/** @typedef {import('./capacity.js').Flag} Flag */

/**
 * Report-performance detectors. Pure: facts in, flags out.
 * @param {{reports?:object[]}} facts
 * @param {object} [config]
 * @returns {Flag[]}
 */
export function detectReports(facts, config = DEFAULT_CONFIG) {
  const reports = facts?.reports ?? [];
  const flags = [];
  for (const r of reports) {
    const where = `${r.workspace} / ${r.name}`;
    if ((r.visuals ?? 0) >= config.report.visualsMin) {
      flags.push({
        type: 'report.too-many-visuals', resource: where, when: '',
        evidence: { visuals: r.visuals },
        what: `Report "${r.name}" has ${r.visuals} visuals on its busiest page.`,
      });
    }
    if (r.mode === 'DirectQuery') {
      flags.push({
        type: 'report.directquery', resource: where, when: '',
        evidence: { source: r.source ?? 'unknown' },
        what: `Report "${r.name}" uses DirectQuery against ${r.source ?? 'an unknown source'}.`,
      });
    }
    if ((r.slowestVisualMs ?? 0) >= config.report.slowVisualMs) {
      flags.push({
        type: 'report.slow-visual', resource: where, when: '',
        evidence: { ms: r.slowestVisualMs },
        what: `Report "${r.name}" has a visual rendering in ${r.slowestVisualMs} ms.`,
      });
    }
  }
  return flags;
}

File: fabric-audit-agent-app/fabric_audit_agent/detectors/report.py
"""Report-performance detectors. Faithful port of the Node ``core/detectors/report.js``."""
from ..config import DEFAULT_CONFIG


def detect_reports(facts, config=None):
    config = config or DEFAULT_CONFIG
    reports = (facts or {}).get("reports") or []
    rep = config["report"]
    flags = []
    for r in reports:
        where = f"{r.get('workspace')} / {r.get('name')}"
        if (r.get("visuals") or 0) >= rep["visualsMin"]:
            flags.append({
                "type": "report.too-many-visuals", "resource": where, "when": "",
                "evidence": {"visuals": r.get("visuals")},
                "what": f"Report \"{r.get('name')}\" has {r.get('visuals')} visuals on its busiest page.",
            })
        if r.get("mode") == "DirectQuery":
            src = r.get("source")   # nullish (?? ), not falsy: an empty-string source is kept
            flags.append({
                "type": "report.directquery", "resource": where, "when": "",
                "evidence": {"source": src if src is not None else "unknown"},
                "what": f"Report \"{r.get('name')}\" uses DirectQuery against {src if src is not None else 'an unknown source'}.",
            })
        if (r.get("slowestVisualMs") or 0) >= rep["slowVisualMs"]:
            flags.append({
                "type": "report.slow-visual", "resource": where, "when": "",
                "evidence": {"ms": r.get("slowestVisualMs")},
                "what": f"Report \"{r.get('name')}\" has a visual rendering in {r.get('slowestVisualMs')} ms.",
            })
    return flags
